### PR TITLE
Build McICA-sampled cloud mask from cloud fraction data

### LINKED ISSUE #301: Build cloud mask for a given cloud fraction

### PR DESCRIPTION
**Pull request by sriharshakandala**
# PULL REQUEST

## Purpose and Content
Builds McICA-sampled cloud mask from cloud fraction data for maximum-random overlap
## Benefits and Risks
(State concisely the benefits to be derived from and the risks associated with this PR)

## Linked Issues
(Provide references to any link issues. Use closes #issuenum to automatically close an open issue)
- Fixes #301 
- Closes #301 

## PR Checklist
- [x] This PR has a corresponding issue OR is linked to an SDI.
- [x] I have followed CliMA's codebase [contribution](https://clima.github.io/ClimateMachine.jl/latest/Contributing/) and [style](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) guidelines OR N/A.
- [x] I have followed CliMA's [documentation policy](https://github.com/CliMA/policies/wiki/Documentation-Policy).
- [x] I have checked all issues and PRs and I certify that this PR does not duplicate an open PR.
- [x] I linted my code on my local machine prior to submission OR N/A.
- [x] Unit tests are included OR N/A.
- [x] Code used in an integration test OR N/A.
- [x] All tests ran successfully on my local machine OR N/A.
- [x] All classes, modules, and function contain docstrings OR N/A.
- [x] Documentation has been added/updated OR N/A.

### CODE REVIEW

**Comment by sriharshakandala** [comment]
bors try

**Comment by sriharshakandala** [comment]
bors try

**Comment by sriharshakandala** [comment]
bors try

File: src/optics/AtmosphericStates.jl [hunk 1]
```diff
@@ -8,12 +8,24 @@ import ..Parameters as RP
...
 using ..Vmrs
 
-export AbstractAtmosphericState, AtmosphericState, GrayAtmosphericState, setup_gray_as_pr_grid, setup_gray_as_alt_grid
+export AbstractAtmosphericState,
+    AtmosphericState,
+    GrayAtmosphericState,
+    setup_gray_as_pr_grid,
+    setup_gray_as_alt_grid,
+    MaxRandomOverlap,
+    ExpRandomOverlap,
```
Comment (szy21) [inline src/optics/AtmosphericStates.jl:17]: Should we remove `ExpRandomOverlap ` since we don't have it now?

File: src/optics/CloudOptics.jl [hunk 2]
```diff
@@ -196,3 +196,65 @@ function pade_eval(ibnd, re, irad, m, n, pade_coeffs, irgh::Union{Int, Nothing}
 
     return (numer / denom)
 end
+
+"""
+    build_cloud_mask!(cld_mask, cld_frac, random_arr, gcol, ::MaxRandomOverlap)
+
+Builds McICA-sampled cloud mask from cloud fraction data for maximum-random overlap
+"""
+function build_cloud_mask!(cld_mask, cld_frac, random_arr, gcol, ::MaxRandomOverlap)
```
Comment (szy21) [inline src/optics/CloudOptics.jl:205]: Do we need to include `random_arr` in the input of this function? Could we define `local_rand` locally?

**Review by szy21 — APPROVED** [review]
Looks good, thanks! I left a few minor comments.

File: src/optics/AtmosphericStates.jl (hunk 1 above)
Comment (sriharshakandala) [reply src/optics/AtmosphericStates.jl]: Done.

File: src/optics/CloudOptics.jl (hunk 2 above)
Comment (sriharshakandala) [reply src/optics/CloudOptics.jl:205]: We are passing a preallocated array, to avoid allocating the random array (of size ngpt) inside this function.

**Pull request merged**